### PR TITLE
Add Android Studio Page

### PR DESCRIPTION
**Pull request by JIceberg**
Nice work Dan 💯

### CODE REVIEW

**Comment by JIceberg** [comment]
#59 
* [x]  Downloading/Installing AS
* [x]  Opening/importing the sdk project
* [ ]  basic overview of the layout (screenshot of Android studio with a project open, highlight the code, the folder on the side, logcat button on the button, terminal button on the bottom, point out run button, clean button, etc)
* [ ]  Gradle
  * [ ]  Troubleshooting
    * gradle clean and rebuild button
    * invalidate cache and restart
  * [ ]  Adding dependencies with gradle?
* [ ]  ADB info
  * [ ]  turning on developer mode and usb debugging, make sure to mention switching the phone away from charge only to the right mode
  * [ ]  adding adb to path
  * [ ]  wireless adb how to
    * [ ]  mention logcat can be used wirelessly
  * [ ]  link to https://developer.android.com/studio/command-line/adb
  * [ ]  talk about keeping config alongside project repo with `adb push` and `adb pull`?
* [ ]  scrpy

**Comment by NoahAndrews** [comment]
> talk about keeping config alongside project repo with adb push and adb pull?

I'd have to look up the exact details, but you can actually include configuration files as an XML resource, so that they get deployed as a part of the APK itself. That's the more correct way to do what you're talking about there, so that you maintain an absolute source of truth.

**Comment by dansman805** [comment]
Oh yeah that's definitely a better way to do it, I just didn't know you could do it that way, that'd be great.

**Comment by NoahAndrews** [comment]
I believe it's as simple as putting a valid configuration xml file in `res/xml`.

**Comment by NoahBres** [comment]
add info for accepting licenses

**Comment by JIceberg** [comment]
Good call. Do you want me to write about how to accept them using the sdk manager GUI or CLI?

**Comment by dansman805** [comment]
> Good call. Do you want me to write about how to accept them using the sdk manager GUI or CLI?

I'd say both if possible.

File: source/docs/software/using-android-studio.rst [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    :alt: Add the dependency using implementation
+
+Finally, perform a gradle sync.
+
+Android Debug Bridge
+====================
+
+Version Control
+===============
+Version control is be extremely useful tool. It allows for looking at (and
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:267]:
```suggestion
Version control is an extremely useful tool. It allows for looking at (and
```

File: source/docs/software/using-android-studio.rst [hunk 2]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
 Android Debug Bridge
 ====================
 
+Logcat
+------
+The logcat is extremely useful for debugging issues
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:269]:
```suggestion
Logcat is extremely useful for debugging issues
```

File: source/docs/software/using-android-studio.rst [hunk 3]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
...
+with your code at runtime or figuring out what went wrong.
+For example, if your app activity crashes and you pull up
+the log seeing 400 pages of something like this:
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:272]:
```suggestion
the log seeing 5000 lines of something like this:
```

File: source/docs/software/using-android-studio.rst [hunk 4]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
...
+
+.. image:: images/using-android-studio/infinite-recursion.png
+    :alt: Infinite recursion
+
+then, you know that there is infinite recursion somewhere
+in your program.
+
+To use the logcat, plug in your device (or connect via ADB).
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:280]:
```suggestion
To use logcat, plug in your device (or connect via ADB).
```

File: source/docs/software/using-android-studio.rst [hunk 5]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
...
+Then, select the app you want to view the logs for. Your window
+should look like this.
+
+.. image:: images/using-android-studio/logcat-window.png
+    :alt: A selected device and app with the error messages
+
+If you have an issue you don't understand, you can print
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:287]: Leaving this here will lead to a *lot* of pointless issues on the SDK repo... maybe suggest the FTC discord? :shrug:

File: source/docs/software/using-android-studio.rst [hunk 6]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
...
+a PDF (option 5) of the log and open an issue on the
+SDK repository.
+
+Wireless Communication
+----------------------
+The Android Debug Bridge (or ADB) is a command-line
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:293]:
```suggestion
Android Debug Bridge (ADB) is a command-line
```

File: source/docs/software/using-android-studio.rst [hunk 7]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
...
+^^^^^^^^^^^^^^
+#.  Ensure USB debugging is enabled on your device and it is
+    in developer mode.
+
+#.  Make sure you have ADB installed. If you do not, follow
+    the instructions at `this link <https://www.xda-developers.com/install-adb-windows-macos-linux/>`_
+
+#.  Connect to the same WiFi network the device is either hosting or on.
+
+Connecting to the Phone
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:307]:
```suggestion
Connecting to a Phone
```

File: source/docs/software/using-android-studio.rst [hunk 8]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
...
+    directory and see if the phone shows up.
+
+#.  Run :code:`adb usb` and then :code:`adb tcpip 5555`. You
+    can then unplug the phone.
+
+#.  Connect to the phone using :code:`adb connect 192.168.49.1:5555`.
+    If this doesn't work, recheck the IP address of the phone and
+    try again with that IP address if it is different.
+
+Connecting to the Control Hub
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:321]:
```suggestion
Connecting to a Control Hub
```

File: source/docs/software/using-android-studio.rst [hunk 9]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Once you're connected to the Control Hub's network,
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:323]:
```suggestion
Once you're connected to a Control Hub's network,
```

File: source/docs/software/using-android-studio.rst [hunk 10]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
...
+you simply need to connect to the device using
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:324]:
```suggestion
you simply need to connect to it using
```

File: source/docs/software/using-android-studio.rst [hunk 11]
```diff
@@ -262,6 +264,69 @@ Finally, perform a gradle sync.
...
+:code:`adb connect 192.168.43.1:5555`.
+
+Once a connection is established, the device should appear
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:327]:
```suggestion
Once a connection is established, it should appear
```

**Review by dansman805 — COMMENTED** [review]
I'd also mention how to add adb to $PATH on windows and probably macOS

File: source/docs/software/using-android-studio.rst (hunk 5 above)
Comment (JIceberg) [reply source/docs/software/using-android-studio.rst:287]: Ah good point. Sometimes issues can be resolved just by having someone who understands error messages read it. Good call, I'll change that.

**Comment by JIceberg** [comment]
I feel like "info on accepting licenses" should be changed to an overall section on the SDK Manager.

File: source/docs/software/using-android-studio.rst [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+
+Downloading the SDK
+-------------------
+The SDK is publicly released to a GitHub
+repository every season. The current season's
+SDK can be found in the
+`FtcRobotController <https://github.com/FIRST-Tech-Challenge/FtcRobotController>`_
+repository.
+
+Downloading the Zip
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:42]:
```suggestion
Downloading the ZIP
```

File: source/docs/software/using-android-studio.rst [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+      :alt: Only select the folder with the Android logo
+
+#. Change to project view. In the top left corner
+   should be a dropdown that allows you to change the
+   way you are looking at your project.
+
+   .. image:: images/using-android-studio/select-project-view.png
+      :alt: Change to project view
+
+#. Wait for `gradle <Gradle>`_ to complete the build.
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:117]:
```suggestion
#. Wait for `Gradle`_ to complete the build.
```
Gradle is a proper noun and should therefore always be capitalized.

File: source/docs/software/using-android-studio.rst [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+.. image:: images/using-android-studio/as-layout.png
+    :alt: Android Studio layout
+    :width: 50em
+
+Creating Classes
+----------------
+The first thing to note in the project view
+is the dropdown with the name of the
+project. If you drop that down, you will
+see all of the gradle files and directories.
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:137]:
```suggestion
see all of the Gradle files and directories.
```

File: source/docs/software/using-android-studio.rst [hunk 15]
```diff
@@ -0,0 +1,426 @@
...
+refer to the official
+`REV documentation <https://docs.revrobotics.com/rev-control-system/>`_.
+
+Gradle
+======
+Gradle is a build tool for software development.
+In the scope of FTC, it is used to build and manage dependencies
+for your project.
+
+When you update any of your gradle files you will need to perform a
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:207]:
```suggestion
When you update any of your Gradle files you will need to perform a
```

File: source/docs/software/using-android-studio.rst [hunk 16]
```diff
@@ -0,0 +1,426 @@
...
+gradle sync, which syncs your project to the changes and rebuilds it.
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:208]:
```suggestion
Gradle sync, which syncs your project to the changes and rebuilds it.
```

File: source/docs/software/using-android-studio.rst [hunk 17]
```diff
@@ -0,0 +1,426 @@
...
+refactoring, etc. The first step is to try cleaning the project
+and doing a rebuild. If this doesn't work, you might have confused
+Android Studio because it caches information about
+your project structure.
+
+The most common way to fix these errors is to do an invalidate
+and restart. In the file dropdown, there will be an option for
+this and then you will choose Invalidate and Restart. This clears
+the cache and restarts your Android Studio, which then should
+perform a gradle rebuild.
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:240]:
```suggestion
perform a Gradle rebuild.
```

File: source/docs/software/using-android-studio.rst [hunk 18]
```diff
@@ -0,0 +1,426 @@
...
+Adding Dependencies
+-------------------
+If you want to add dependencies to your project, you can do so
+in the :code:`build.gradle` file in the TeamCode directory.
+
+There should be a dependencies block at the bottom of the file.
+
+.. image:: images/using-android-studio/teamcode-gradle.png
+
+Some dependencies require changes to your other gradle files.
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:251]:
```suggestion
Some dependencies require changes to your other Gradle files.
```

File: source/docs/software/using-android-studio.rst [hunk 19]
```diff
@@ -0,0 +1,426 @@
...
+Make sure to read the installation instructions for whatever
+gradle dependency you want to add.
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:253]:
```suggestion
dependency you want to add.
```
The use of "gradle" is redundant.

File: source/docs/software/using-android-studio.rst [hunk 20]
```diff
@@ -0,0 +1,426 @@
...
+      :alt: Change to project view
+
+#. Wait for `gradle <Gradle>`_ to complete the build.
+
+   .. image:: images/using-android-studio/build-gradle.png
+      :width: 50em
+
+Layout
+======
+Android Studio can look menacing at first glance, but
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:124]:
```suggestion
Android Studio can look intimidating at first glance, but
```
It's just a more fitting word imo :shrug:

File: source/docs/software/using-android-studio.rst [hunk 21]
```diff
@@ -0,0 +1,426 @@
...
+
+.. note::
+    You can use logcat via ADB with the
+    :code:`adb logcat` command. This is extremely useful for
+    debugging as it allows you to look at the logs wirelessly
+    which saves time. Remember, logcat is the *best* way
+    to debug your software.
+
+Connecting to a Phone
+^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:328]:
```suggestion
Connecting to a Phone Wirelessly
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: source/docs/software/using-android-studio.rst [hunk 22]
```diff
@@ -0,0 +1,426 @@
...
+
+#.  Run :code:`adb usb` and then :code:`adb tcpip 5555`. You
+    can then unplug the phone.
+
+#.  Connect to the phone using :code:`adb connect 192.168.49.1:5555`.
+    If this doesn't work, recheck the IP address of the phone and
+    try again with that IP address if it is different.
+
+Connecting to a Control Hub
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (dansman805) [inline source/docs/software/using-android-studio.rst:342]:
```suggestion
Connecting to a Control Hub Wirelessly
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Pull request merged**